src/idpyoidc/server/session/grant_manager.py

```diff
@@ -0,0 +1,292 @@
+"""
+Implements a database with branches with N nodes (SessionInstance instances)
+that ends in a Grant instances.
+"""
+import logging
+from typing import Callable
+from typing import List
+from typing import Optional
+from typing import Union
+
+from idpyoidc.encrypter import default_crypt_config
+from idpyoidc.encrypter import get_crypt_config
+from idpyoidc.message.oauth2 import TokenExchangeRequest
+from idpyoidc.server.session.info import ClientSessionInfo
+from idpyoidc.server.token import handler
+from .database import Database
+from .grant import ExchangeGrant
+from .grant import Grant
+from .info import NodeInfo
+from ..exception import InvalidBranchID
+from ..token.handler import TokenHandler
+
+logger = logging.getLogger(__name__)
+
+
+class GrantManager(Database):
+    parameter = Database.parameter.copy()
+    init_args = ["handler"]
+
+    def __init__(
+            self,
+            handler: TokenHandler,
+            conf: Optional[dict] = None,
+            remember_token: Optional[Callable] = None,
+            remove_inactive_token: Optional[bool] = False,
+    ):
+        self.conf = conf or {
+            "session_params": {
+                "encrypter": default_crypt_config(),
+                "node_type": ["client", "grant"],
+                "node_info_class": {"client": ClientSessionInfo, "grant": Grant}
+            }
+        }
+
+        session_params = self.conf.get("session_params") or {}
+        _crypt_config = get_crypt_config(session_params)
+        super(GrantManager, self).__init__(_crypt_config, **self.conf)
+
+        self.token_handler = handler
+        self.remember_token = remember_token
+        self.remove_inactive_token = remove_inactive_token
+
+    def get_salt(self):
+        """returns the original salt assigned in init"""
+        return self.crypt_config["kwargs"]["salt"]
+
+    def __setattr__(self, key, value):
+        if key in ("_key", "_salt"):
+            if hasattr(self, key):
+                # not first time we configure it!
+                raise AttributeError(f"{key} is a ReadOnly attribute that can't be overwritten!")
+        super().__setattr__(key, value)
+
+    def __getitem__(self, branch_id: str):
+        return self.get(self.decrypt_branch_id(branch_id))
+
+    def __setitem__(self, branch_id: str, value):
+        return self.set(self.decrypt_branch_id(branch_id), value)
+
+    def _setup_branch(self, path):
+        for i in range(len(path)):
+            _id = path[0:i + 1]
+
+            try:
+                _si = self.get(_id)
+            except KeyError:
+                _info_class = self.node_info_class[self.node_type[i]]
+                _si = _info_class()
```

Comment:
Should we pass the id here?

`_si = _info_class(_id[i])`

Reply:
Yes, makes sense!
Good catch!